### ISSUE TITLE
Crash [TEALExceptionHelper checkAndSendCrashLog]

### ISSUE THREAD

**Issue by Tokanod** [Tealium/ios-library#28]
Hi,

We recently updated our Tealium version to 4.1.6.
Now, I can see a lot of crashes related to that in this version:

```
Thread : Crashed: com.tealium.shared-thread
0  libsystem_kernel.dylib         0x3902f1f0 __pthread_kill + 8
1  libsystem_pthread.dylib        0x39099797 pthread_kill + 58
2  libsystem_c.dylib              0x38fdffdd abort + 76
3  libc++abi.dylib                0x3830ecd7 abort_message + 74
4  libc++abi.dylib                0x383276e5 default_terminate_handler() + 252
5  libobjc.A.dylib                0x38a6cf65 _objc_terminate() + 192
6  libc++abi.dylib                0x383251c7 std::__terminate(void (*)()) + 78
7  libc++abi.dylib                0x38324a1d __cxxabiv1::exception_cleanup_func(_Unwind_Reason_Code, _Unwind_Exception*)
8  libobjc.A.dylib                0x38a6cda3 objc_exception_throw + 250
9  CoreFoundation                 0x2dea6967 -[NSObject(NSObject) doesNotRecognizeSelector:] + 202
10 CoreFoundation                 0x2dea5253 ___forwarding___ + 706
11 CoreFoundation                 0x2ddf47b8 _CF_forwarding_prep_0 + 24
12 Foundation                     0x2e81d5a7 -[NSKeyedUnarchiver initForReadingWithData:] + 310
13 Foundation                     0x2e81d401 +[NSKeyedUnarchiver unarchiveObjectWithData:] + 52
14 MyApp                         0x00584fab +[TEALExceptionHelper checkAndSendCrashLog] (TEALExceptionHelper.m:61)
15 MyApp                         0x00586c9f +[TealiumActual initWithCompletionBlock:] (TealiumActual.m:155)
16 MyApp                         0x0058606d +[TealiumActual initSharedInstance:profile:target:options:globalCustomData:completionBlock:] (TealiumActual.m:87)
17 MyApp                         0x0056b899 __70+[Tealium initSharedInstance:profile:target:options:globalCustomData:]_block_invoke (Tealium.m:30)
18 libdispatch.dylib              0x38f54833 _dispatch_call_block_and_release + 10
19 libdispatch.dylib              0x38f5af11 _dispatch_queue_drain$VARIANT$mp + 488
20 libdispatch.dylib              0x38f5ac97 _dispatch_queue_invoke$VARIANT$mp + 42
21 libdispatch.dylib              0x38f5ba45 _dispatch_root_queue_drain + 76
22 libdispatch.dylib              0x38f5bd29 _dispatch_worker_thread2 + 56
23 libsystem_pthread.dylib        0x39096bd3 _pthread_wqthread + 298
```

A big part of the crashes seem to happen on iOS 7, but we do have a part on iOS 8 too (we are supporting only those two versions).

Any help with that would be much appreciated  :)

Thank you.
Audrey

**Comment by ethyreal** [comment]
Hi Audrey,

Thank you for the feedback! 

This is an unfortunate problem.  What is happening is with exception logging enabled the Tealium Library attempts to capture any exceptions thrown and archive them in order to dispatch them later after the next successful launch.  This works fine unless an exception didn't finish archiving.  The system gives your application very little time in which to process any exception handling so it is possible for the archive to not complete.  As you can see in step 13 of the stack trace you posed:
```
13 Foundation                     0x2e81d401 +[NSKeyedUnarchiver unarchiveObjectWithData:] + 52
```
unarchiveObjectWithData is throwing an exception.  This is one of the only places in Cocoa's public API's that actually does throw exceptions.  While normally in Cocoa you never want to catch exceptions because they are fatal to the system and can have unforeseen consequences so you actually do want to crash, in this case we can safely catch the exception.

So in short, we have pushed up a patch fixing this issue which you can download below or just pull from our master branch if you have the repo cloned:

https://github.com/Tealium/ios-library/releases/tag/4.1.9

Kind regards,

George

**Comment by Tokanod** [comment]
Hi George,

Thank you for your quick answer. That is very helpful.
We will update the library and I'll keep you in touch in the next few weeks (we'll ship that in the next release in 2 weeks probably).

Best regards,
Audrey
